ghproxy/ghcache/coalesce.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ghcache
+
+import (
+	"bufio"
+	"bytes"
+	"net/http"
+	"net/http/httputil"
+	"sync"
+)
+
+// requestCoalescer allows concurrent requests for the same key to share a
```

Comment:
how often do we expect this to occur?

Reply:
what is the RTT like? this still feels like something the cache should handle, without needing a global lock or any of the other complexity. this feels like purely premature optimization. I foresee hung goroutines.

Reply:
RTT is between 0.25s and 0.60s from the Prow cluster.
I agree that this is something the cache should handle. The vendored caching layer doesn't handle it itself so I wrapped it with a transport so that the ghcache handles it transparently.

I don't think this is a premature optimization. This is a common case for the most used API calls like `GetIssueLabels` and `ListIssueComments`. The logs show that there are frequently multiple calls for the same resource occurring within a second (the time granularity that stackdriver displays). Here is a sample showing just the `GetIssueLabels` calls for an event:
```
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info","component":"hook"} 
2018-04-24 17:12:15.000 PDT  {"client":"github","level":"info","component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
2018-04-24 17:12:15.000 PDT  {"component":"hook","msg":"GetIssueLabels(kubernetes, kubernetes, 62297)","client":"github","level":"info"} 
```
Without the synchronization layer I added, all of the above calls would use an API token if the cache is stale instead of just using one token in total.

I'm not too worried about the global lock. We don't hold it while executing requests, only to lookup if a request is in flight and add/delete entries. So requests for different paths won't block each other for more than the time it takes to update the map (which is negligible when compared to RTT time to Github).
I can change the map to be a concurrent `sync.Map` if that is preferable though.

Reply:
> I'm not too worried about the global lock. We don't hold it while executing requests, only to lookup if a request is in flight and add/delete entries. So requests for different paths won't block each other for more than the time it takes to update the map (which is negligible when compared to RTT time to Github).

I'm much more concerned with all the waiters etc. That code looks ripe for bugs / hanging.

> I can change the map to be a concurrent sync.Map if that is preferable though.

sync.Map is not suited to this. it's a rather complex map internally only optimized for specific use cases (maps that are append-only and read heavy, or also maps that are only a ~ fixed set of keys where only the values are touched, for some patterns)

Reply:
similarly https://github.com/kubernetes/test-infra/pull/7848/files#r183921856 it looks to me like the keys map in requestCoalescer is append-only right now but probably shouldn't be..

Reply:
I added a unit test. Let me know what you think.